### PR TITLE
store/tikv: pre-split regions during 2PC to avoid hotspot

### PR DESCRIPTION
**Pull request by tiancaiamao**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/16573

Problem Summary:

Pre-split region before it become a writing hotspot to avoid the TiKV 'server is busy' error.

### What is changed and how it works?

What's Changed:

Pre-split the region during 2PC if we detect there are many mutations into one region.

How it Works:

Call the `store.SplitRegions` and `store.WaitScatterRegionFinish` API.

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Manual test (add detailed scripts or steps below)

http://172.16.4.61:13000/d/000000011/test-cluster-tidb?orgId=1&from=1588073060157&to=1588082547757

### CODE REVIEW

**Comment by tiancaiamao** [comment]
PTAL @youjiali1995 @lysu @cfzjywxk @jackysp

File: store/tikv/2pc.go [hunk 1]
```diff
@@ -435,6 +439,66 @@ func (c *twoPhaseCommitter) doActionOnMutations(bo *Backoffer, action twoPhaseCo
...
+	}
+
+	regionIDs, err := store.SplitRegions(ctx, splitKeys, true)
+	if err != nil {
+		logutil.BgLogger().Warn("2PC split regions failed", zap.Uint64("regionID", group.region.id), zap.Int("keys count", length), zap.Error(err))
+		return false
+	}
+
+	for _, regionID := range regionIDs {
+		err := store.WaitScatterRegionFinish(regionID, 0)
```
Comment (lysu) [inline store/tikv/2pc.go:491]: do we have problem when sqlA wait on here, but sqlB try to resplit same region to smaller one again?

File: store/tikv/2pc.go [hunk 2]
```diff
@@ -435,6 +439,66 @@ func (c *twoPhaseCommitter) doActionOnMutations(bo *Backoffer, action twoPhaseCo
 		return errors.Trace(err)
 	}
 
+	// Pre-split regions to avoid too much write workload into a single region.
+	// In the large transaction case, this operation is important to avoid TiKV 'server is busy' error.
+	var preSplited bool
+	for _, group := range groups {
+		if group.mutations.len() > preSplitThreshold || testSplitRegionFlag {
```
Comment (coocood) [inline store/tikv/2pc.go:446]: I think it's better to make preSplitThreshold a variable and modify it in the test.

File: store/tikv/2pc.go [hunk 3]
```diff
@@ -435,6 +439,66 @@ func (c *twoPhaseCommitter) doActionOnMutations(bo *Backoffer, action twoPhaseCo
...
+
+func preSplitAndScatterIn2PC(ctx context.Context, store *tikvStore, group groupedMutations) bool {
+	length := group.mutations.len()
+	splitKeys := make([][]byte, 0, 4)
+
+	const sz32M = (32 << 20)
+	regionSize := 0
+	for i := 0; i < length; i++ {
+		regionSize = regionSize + len(group.mutations.keys[i]) + len(group.mutations.values[i])
+		if regionSize >= sz32M || (testSplitRegionFlag && i == length/2) {
```
Comment (coocood) [inline store/tikv/2pc.go:475]: Better make `sz32M` a variable and modify it in the test.

File: store/tikv/2pc.go (hunk 1 above)
Comment (tiancaiamao) [reply store/tikv/2pc.go:492]: It doesn't affect the final result. `WaitScatterRegionFinish` has its own timeout, and even the operation fail, the 2PC process continues.

There shouldn't be many large transactions modify the same region (and split it), so what you concerned is unlikely to happen.

**Pull request merged**